src/util/psbt/map/output.rs

```diff
@@ -55,6 +73,41 @@ pub struct Output {
     pub unknown: BTreeMap<raw::Key, Vec<u8>>,
 }
 
+/// Taproot Tree representing a finalized [`TaprootBuilder`]
+#[derive(Clone, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct TapTree(pub(crate) TaprootBuilder);
+
+impl PartialEq for TapTree {
+    fn eq(&self, other: &Self) -> bool {
+        // Invariant: builder is finalized
+        self.node_info().hash.eq(&other.node_info().hash)
+    }
+}
+
+impl TapTree {
+
+    // get the inner node info as the builder is finalized
+    fn node_info(&self) -> &NodeInfo {
+        self.0.branch[0].as_ref().unwrap()
+    }
+
+    /// Convert a [`TaprootBuilder`] into a tree if it is complete binary tree.
+    pub fn from_inner(inner: TaprootBuilder) -> Option<Self> {
+        if inner.is_complete() {
+            Some(TapTree(inner))
+        } else {
+            None
+        }
+    }
```

Comment:
AFAIK it's a [convention](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.from_utf8) in Rust to not destroy data like this. `Result<Self, TaprootBuilder>` would be better.

Reply:
Good catch. Did not know about this convention